### ISSUE TITLE
Using JOML types as constants/immutability

### ISSUE THREAD

**Issue by immortius** [JOML-CI/JOML#52]
I heard about JOML today, and as I'm always on the look out for a good math library have been looking over it.

I guess my one concern (and this was an issue with vecmath too) is that there is no safe way to have a constant of any of the JOML types. Because all the class variables are public they are always mutable. So something like:

``` java
public static final Vector3f UP = new Vector3f(0, 1, 0);
```

Is always at risk of unintentional (or malicious) modification. (public variables also prevent injecting byte code to intercept value changes as that requires the use of setters, although that is not an issue for me as such).

Possible approaches to address, in order of ascending complexity:
### 1. Completely separate immutable types

This could be done outside of JOML - having a separate class of each type that is immutable, and can produce the JOML class as needed.

Downside is object allocation.
### 2. Make class variables private, add accessor methods

something like

``` java
private float x, y, z;

public float x() {
  return x;
}

public void x(float value) {
  x = value;
}
```

This would allow a subclass to prevent changes by throwing exceptions on usage of method that would modify the class. The subclasses would still be usable in expected ways though such as `v.add(UP)`.

Downside is possibility of accidental leakage of immutable types into mutable context, resulting in a crash down the track when an attempt is made to mutate them. At least that is an obvious issues as opposed to the accidental alteration of a constant which can be hard to debug.
### 3. Read-only base interface

A 'trick' which I came up with that was used by the [TeraMath](https://github.com/MovingBlocks/TeraMath) library:

First, you have an interface that provides the non-modifying methods of the type:

``` java
public interface Vector3fConst {
   float x();
   float y();
   float z();

   float length();
}
```

(this could also be an abstract class, with methods like length implemented in terms of the variable accessor methods)

then you have the mutable implementation

``` java
public final class Vector3f implements Vector3fConst {
   public float x, y, z;

   public Vector3f set(Vector3fConst other) {
      x = other.x();
      y = other.y();
      z = other.z();
      return this;
   }  
}
```

This provides a few benefits:
- Method signatures clearly indicate whether they mutate a vector argument. If you send a Vector3f to the set method, you know it will not be modified.
- It is possible to implement a true immutable vector to guarantee that the user won't just cast the value to the mutable implementation. This won't have a bunch of mutating methods that throw exceptions.
- Also helps prevent accidental sharing of values between contexts. If you have the method

``` java
void setEntityPosition(Vector3f newPosition)
```

There is always the worry that it does

``` java
position = newPosition;
```

instead of the desired

``` java
position.set(newPosition);
```

With the signature

``` java
void setEntityPosition(Vector3fConst other);
```

you know even if it is shared it won't be modified at least.

Downside is the method calls required access the values, the extra complexity of the class model, and without the actual immutable type it is only a soft guarantee of immutability. Also it is sort of dragging C++ behaviors into Java, which may be confusing.

Anyway, just wanted to share these thoughts and ideas.  I don't know whether these sorts of concerns are within the scope of JOML or not.

**Comment by httpdigest** [comment]
Thanks for your post. Immutability of data structures is certainly an interesting aspect.

Generally, I agree with you on the aspect that mutability of data structures allows to change them in unexpected ways which in turn can introduce errors in an application.
I would even go so far as to denunciate this as "bad design."
And I have become very aware of that while in the past I have been struggling with customers with their own frameworks that had exactly this (mutability of core data structures) as their main problem, which in turn increased the rigidity and fragility of the whole code base and made any changes whatsoever literally impossible to do.
There are two forces that drove the design in JOML to use mutable data structures:
- readability of code: Using vector.x = 3 is the mosts concise and understandable way of assining the x-component of a vector to 3. Nothing can beat that in readability
- reduction of garbage and not relying on escape analysis: I know that mature desktop Java Runtimes have the ability to see whether an allocated object does not escape a certain scope and then avoid allocation altogether. This however cannot be relied upon in other execution environments, such as Android

Currently, I see only one way for an application to provide immutability of matrix/vector/quaternion classes. And that is by copying the API of the respective JOML class and delegating to a JOML instance.
That is, currently I would not want to have JOML take the responsibility of reducing rigidity and fragility of client-side code by making JOML's classes immutable. That is a responsibility which I see on the side of the client using JOML.
However, since people are starting to request an interface-style API of JOML (see https://github.com/JOML-CI/JOML/issues/51), this would make providing immutable data types easier.
I'll think about that.

**Comment by immortius** [comment]
Please note that I am certainly not suggesting that JOML classes should be immutable - I absolutely agree with the two concerns you raise.   I am merely interested in approaches for allowing support for immutable instances (or at least instances with some guarantee of not being changed as per suggestion 3) that can interop with JOML types. Preferably without having to instantiate a JOML object on each use.

**Comment by roquendm** [comment]
KISS:

public class CVector3f {  
  public final float x,y,z;
  public CVector3f(float x, float y, float z)  {    this.x=x; this.y=y; this.z=z;  }
}

and a setting method to Vector3f and you're done.

**Comment by httpdigest** [comment]
@immortius: Now that I've read through your point 3 again, I began to realize what the actual problem is. And I have been having this problem too with javax.vecmath in the past, I remember.

So, the essence of the problem is just "object identity," and that Java does not (yet) have "value types."
We want to avoid object allocations and therefore share objects using multiple references to them, but we still want to treat vectors as value types without an identity.

This is an inherent problem in the language/runtime and I guess it cannot be solved without altering the language itself. Other cases where this exact problem occurs (sharing to avoid memory allocations, but still treating as separate values) use "copy-on-write semantics."

But this necessitates to have a single instance for each "handle" object, at least, and delegate to a shared JOML vector object; and once any modification is done on it, create a copy of it.
This would however not save us much (only the fields of the vector). And it would make everything a hell more complicated and probably even less efficient because of the JVM not optimizing as aggressively as it could.

**Comment by immortius** [comment]
I guess I agree that ideally math types would be value types, which Java doesn't support (without customising the JVM, using Scala ? or other shenanigans at least). So any approach taken will make sacrifices in some area. However I think the gap can be closed a bit without having to do something major like switch to immutable-only types or a copy-on-write approach.

I hadn't even thought of copy-on-write semantics, and I imagine that it would represent a pretty severe change in API if nothing else. I think the standard, mutable JOML objects and their API are fine - and are actually excellent in the majority of situations. I assume you wouldn't want to make large scale breaking changes to the API. In that regard approach 2 fails, since it requires getting rid of all the public variables.

I think the advantage of approach 3 is that the API for the existing JOML objects would remain the same, except 

1. they would inherit an interface which defines just the non-mutating methods.
2. some of their methods would change to take instances of the interfaces instead of the actual objects, where they don't need to mutate them.
3. they gain an accessor method for each parameter

But they can keep their public variables (thus maintaining the readability benefit), and they can continue to be used the same way as previously. But the interface allows an alternative implementation with no modification allowed, which can still be passed to operations on the standard implementation. The alternative implementation would not itself have any mutating operations.

If it would help I could PR an example of what I mean? Or if you are not interested that is also fine.

**Comment by httpdigest** [comment]
Thanks for the discussion so far with your elaborate explanations, and for your offer of doing a pull request!
I would very much appreciate a pull request on this and created a new branch "immutable" (based off the current master branch) which you can target for a PR.

**Comment by httpdigest** [comment]
I am closing this for now. If more people ask for immutability in JOML we could do an official build/variant of the [immutable](https://github.com/JOML-CI/JOML/tree/immutable) branch.
But as always, commits are always welcome, especially for keeping `immutable` in sync with `master`.

**Comment by immortius** [comment]
So it has been a year since this was done, so I guess it is worth having a little postmortem.

- I didn't keep the branch up to date. I did look at updating it in march, but merging is impossible given the structural changes and manually applying updates was laborious.
- The branch was never released.
- Having a branch doesn't really achieve having a single, standard math library anyway, although it is one-way source compatible with master (can move from master to immutable without breaking source).

I would like to keep this alive, but it doesn't seem worth the effort if it isn't going to be released. I guess a couple of options for making maintaining it a bit easier are:
 - Change master to have at least the constant/mutable class split (dropping the immutables, they're not important anyway). 
 - Change master to generate from string templates or use some other code generation technique, similar to [TeraMath](https://github.com/MovingBlocks/TeraMath). Different build options could be used to generate as single classes or generate the split constant/mutable classes. This may also provide some further advantage in generating float/double versions of the same class from the same template, and possibly even different dimensioned versions. The disadvantage is that the code will be a bit more difficult to maintain (since you have to update templates rather than source code).
 - Maybe pursue some trickiness to convert the source from master into source for the branch.

Otherwise I guess I either need to give up on the safety offered by the constant/mutable split, or very regularly maintain the fork.

On the bright side, in making joml-immutable a couple of bugs were discovered and fixed in master - a couple of methods were altering parameters they should not have been, and a couple of methods were returning their owning object rather than a new object. This does highlight the advantage of the constant/mutable split - a method declares which parameters will be mutated, and methods on the constant class cannot mutate the owning instance. With the split the bugged methods would not compile, so such bugs are easily detected.

**Comment by httpdigest** [comment]
Hey @immortius,

I thought about it and about the possible ways to accomplish a maintainable solution which is also easy to implement (without touching any of the existing classes). For this I would like to ask you first, whether it is more important for you to have compile-time typesafety to check whether someone accesses mutating methods through an immutable interface; _or_ whether the interface of a Matrix4f can remain as is, and validation is deferred until runtime via some kind of "ObjectNotMutableException" (or the like).

**Comment by httpdigest** [comment]
I opted for the following solution now which is ready for Matrix4f:
- provide a `toImmutable()` method which instantiates an "immutable" version of that class (with the same static interface, so no new "Immutable" classes/interfaces are introduced)
- the returned object delegates all _non-mutating_ methods to the object from which it was created
- it throws a runtime exception whenever a mutating method was invoked which would result in `this` being modified

The problem here lies within the last point: Not all methods that are potentially mutating do mutate `this`. This is true for all methods whose result is a function of `this` and another parameter. In those cases, an additional `destination` parameter is supplied. This _can_ be `this`. An example is `translate(Vector3f, Matrix4f)`.

The current solution relies on a generic algorithm that generates an immutable class on demand at runtime, which then extends the class for which an immutable object should be created.
All methods for which it can statically be proven that they mutate `this` will be overwritten in the immutable class and an immediate runtime exception will be thrown.
For methods with a `dest` parameter (which can be something other than `this`) a check is made to see whether `dest == this`, in which case an exception is thrown.

All other methods that only perform reads, are not overwritten, so virtually delegate to the delegate object.

**Comment by immortius** [comment]
I would say my main interest is in the communication and to some extent enforcement of when a math value type is invariant when used by a method, to make it clearer how to correctly and efficiently use a method. I don't actually care about actual immutablity, and I regret ever mentioning immutables at this point - I think it has confused things greatly. 

To give an example, when faced with a method like this:

```
class Player {
    public void setLocation(Vector3f newLocation)
}
```

Behind the scenes it may either do:
``` this.location = newLocation; ```
or
```this.location.set(newLocation)```

If I'm trying to teleport the player to a beacon, doing:
``` player.setLocation(beacon.getLocation())```
Is either the correct and efficient way, or will entangle the locations of player and beacon such that they move around together. Alternatively:
``` player.setLocation(new Vector3f(beacon.getLocation());```
Is either correct, or an inefficient usage that results in an extra object allocation.

The set example is perhaps a little weak, since in general you are better off just having a getter and working with the player's location directly, but this is true for any method that works with the math types.

The solution I had proposed, with an underlying invariant interface, provides a way to clarify this - if the method is
```void setLocation(Vector3f newLoc)```
then it means the method may result in the provided object being altered - so you should create a defensive copy if you don't want that to happen. While
```void setLocation(Vector3fi newLoc)```
communicates that the provided object will not be altered - so you can pass an existing object safely to the method. This is somewhat enforced by the underlying Vector3fi not having any mutable methods declared (the method could cast around this though).

The toImmutable() method doesn't really address this. I guess a client to a method can create immutable versions of every argument they don't expect to be changed, but this isn't very efficient. But I'm pretty sure you are not suggesting they would be used this way anyhow.

...

All that aside, this seems like a reasonable way to implement Immutables - it mirrors the Collections.unmodifiableX() methods, and keeps them simple. They code could potentially be generated via annotation processors or similar even. Although... this technique won't work for the Vectors and similar with their public fields.

**Comment by httpdigest** [comment]
> They code could potentially be generated via annotation processors or similar even.

I think, the best way to handle all of this would indeed be via annotation processors, like the [Checker Framework](http://types.cs.washington.edu/checker-framework/) and its support for [immutable](http://types.cs.washington.edu/checker-framework/current/checker-framework-manual.html#igj-checker) references.

**Comment by httpdigest** [comment]
I gave it a try and introduced a Vector3fc (for Vector3f 'const'), which only contains non-mutating methods, and adapted the rest of the classes to it. It feels really nice to statically know when the vector is _not_ going to be modified by a method to which it is applied, and to statically get a compiletime error when applying the vector to a method that only takes the mutable class type.
I think I am going to stay with this.

**Comment by immortius** [comment]
Mmm, this is similar to what I did in the original immutables branch, except I had Vector3fc as an abstract class with the non-mutating methods in it. I did that primarily so the immutable implementations could inherit them. In retrospect having Vector3fc as an interface is a much cleaner solution. The immutable variation can still be a class (probably generated) that implements the interface, but then delegating its methods to a wrapped Vector3f. Then there is no need to move the implementations around. 

The slight loss in not doing the split is the methods declared in Vector3fc could still be implemented as mutating - if they are defined in Vector3fc then they don't have access to mutating elements. But this shouldn't be a big issue since it only affects the library itself, not client code. Could also be picked up with static code analysis.

I'm very happy with this solution. 👍

**Comment by httpdigest** [comment]
Thanks for your feedback!
I will merge the current solution into master, then, and start working on the immutable versions of the remaining classes.

As I see it, we now have two ways to make an object immutable:
1. by providing static typesafety with an interface containing only the immutable methods (currently implemented for Vector3f via Vector3fc interface)
2. by creating a subclass (currently at runtime) which enforces immutability in mutating methods by throwing an exception (currently implemented for Matrix4f via Matrix4f.toImmutable()

Of these two methods, I actually now prefer the first, and will drop the second.

> The immutable variation can still be a class (probably generated) that implements the interface, but then delegating its methods to a wrapped Vector3f. Then there is no need to move the implementations around.

Yes, that would certainly also be possible, and would even disallow clients to cast Vector3fc to the mutable Vector3f class if they should not be allowed to.

**Comment by httpdigest** [comment]
Immutability for the following selected classes has been implemented now (due to priority of their needs):
- Vector3f, Vector3d, Vector4f, Quaternionf, Matrix4f, Matrix4d

There are two ways to create an immutable view on such instances:
- via exposing the instance through their immutable interface
- via the toImmutable() instance method on each such class

The latter instantiates a delegating proxy class to disallow clients to cast an object from its immutable interface to the implementation class.
The solution does not need any additional dependencies anymore and is again 1.4 compatible.

The current version can be obtained via Maven/Gradle using the coordinates org.joml:joml:1.9.0-SNAPSHOT or via the GitHub repo.

**Comment by isaac-udy-8i** [comment]
Why does the Android build of JOML not include the immutable classes?

**Comment by httpdigest** [comment]
Because Android cannot optimize interface method calls and JOML (particularly the Android version which is also used in Samsung's GearVR Framework) has to be geared towards peak performance.

The HotSpot JVM on the other hand can eliminate interface calls when there is only 1 class loaded implementing that interface. This is why interface calls of those "immutable" (btw. now "read-only") methods can be accelerated and in most cases via inlining completely eliminated.
